### PR TITLE
Fix ruby time support

### PR DESCRIPTION
**Pull request by saks**
When using a ruby adapter we noticed an issue with 1 second difference. I was able to reproduce the issue with tests but had to add more methods into `chronological.udl`. Please let me know if you have any questions.

### CODE REVIEW

File: uniffi_bindgen/src/bindings/ruby/templates/RustBufferBuilder.rb [hunk 1]
```diff
@@ -114,7 +114,11 @@ def write_{{ canonical_type_name }}(v)
     nanoseconds = v.tv_nsec
 
     if seconds < 0
+      # In order to get duration nsec we shift by 1 second:
```
Comment (bendk) [inline uniffi_bindgen/src/bindings/ruby/templates/RustBufferBuilder.rb:117]: I think this comment can be improved to describe this shift a bit more precisely.  IIUC, the issue is that Ruby and UniFFI represent times before the epoch differently.  For Ruby, the nanoseconds value is it's own positive value that should always be added to the time.  For UniFFI, the nanoseconds value is part of the number, so it should be subtracted when the seconds are negative.   BTW, I'd guess that more thought was put in to the Ruby representation and wonder if it has advantages over ours.

Also, I think this shift should be skipped when `nanoseconds=0`.  Since in that case you're going to end up with `nanoseconds=10^9`, which overflows it's bounds.

**Review by bendk — CHANGES_REQUESTED** [review]
The changes look good, but I'm hoping you can update the comments a bit and also fix a corner case.

File: uniffi_bindgen/src/bindings/ruby/templates/RustBufferBuilder.rb (hunk 1 above)
Comment (saks) [reply uniffi_bindgen/src/bindings/ruby/templates/RustBufferBuilder.rb:122]: Thanks for looking into this @bendk. Indeed ruby implementation is weird and the main reason is the lack of duration type/class in the standard library. One may think that `Time` can be used as a replacement, but unfortunately ruby doesn't support math operations between two time objects:
```ruby
[5] pry(main)> Time.at(0) + Time.at(1)
TypeError: time + time?
from (pry):5:in `+'
```
Hence I had to figure out a way to convert properties of time object into values that ffi type conversion layer can interpret correctly. I've added comments clarifying this appraoch. Please let me know if these are helpful.

I appreciate your point about the corner case with `nanoseconds=0`. No shift is necessary for that scenario, I've updated the code. One issue though is that rust code seems handles overflow gracefully so I can't find a way to reproduce it in tests.

**Review by bendk — APPROVED** [review]
The comments look great, thanks for updating them!
